### PR TITLE
[node-local-dns] New health checking machinery

### PR DESCRIPTION
**Pull request by zuzzas**
## Description

Rewrote ready checking. Removed locking. All logic resides in the sidecar container now.

## Why we need it and what problem does it solve?

1. iptables rules gets removed within one second of coredns being shutdown
2. if a coredns process closes a listening socket (typically due to unexpected death of the coredns), `-m socket` will stop matching

## Changelog entries

```changes
module: node-local-dns
type: fix
description: Reworked health checking logic
note: Now Pods shouldn't crash unexpectedly now due to poor implementation of locking/probing.
```

### CODE REVIEW

**Review by konstantin-axenov — CHANGES_REQUESTED** [review]

File: ee/modules/350-node-local-dns/templates/daemonset.yaml [hunk 1]
```diff
@@ -65,20 +65,12 @@ spec:
         - containerPort: 53
           name: dns-tcp
           protocol: TCP
-        livenessProbe:
+        readinessProbe:
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply ee/modules/350-node-local-dns/templates/daemonset.yaml:74]: First part is done.

I do not believe that it is reasonable to restart the container earlier than one minute in its lifetime. We install no iptables rules and gain nothing by constantly restarting the container.

File: ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+function add_rule() {
+  if ! iptables -w 600 -W 100000 -t raw -C PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK >/dev/null 2>&1 ; then
+    iptables -w 600 -W 100000 -t raw -A PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK
+  fi
+}
+
+function delete_rule() {
+  if iptables -w 600 -W 100000 -t raw -C PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK >/dev/null 2>&1 ; then
+    iptables -w 600 -W 100000 -t raw -D PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh]: Done.

File: ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+pipe=$(mktemp -u)
+mkfifo "$pipe"
+exec 3<>"$pipe"
+rm "$pipe"
+
+echo "first run" >&3
+
+inotifywait -q -m -e modify "$readiness_file_path" >&3 |
+while read -r filename event <&3; do
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh]: Done.

File: ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      delete_rule
+      latest_state="$not_ready_state"
+    fi
+  else
+    echo "Unknown state in file \"$readiness_file_path\": \"$(< "$readiness_file_path")\""
+    exit 1
+  fi
+}
+
+trap delete_rule INT TERM
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply ee/modules/350-node-local-dns/images/iptables-loop/iptables-loop.sh]: Done.

File: ee/modules/350-node-local-dns/images/coredns/readiness.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+# Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+
+set -Eeuo pipefail
+
+function error() {
+  echo -n "not-ready" > /tmp/coredns-readiness
+  exit 1
+}
+
+trap error ERR
```
Comment (zuzzas) [inline ee/modules/350-node-local-dns/images/coredns/readiness.sh:13]: EXIT
```suggestion
trap error ERR EXIT
```

**Review by konstantin-axenov — APPROVED** [review]

File: ee/modules/350-node-local-dns/images/coredns/start.sh [hunk 6]
```diff
@@ -16,15 +18,15 @@ if ! ip -json addr show "$dev_name" | jq -re "any(.[].addr_info[]?.local; . == \
...
+if ! iptables -w 60 -t raw -C OUTPUT -s "${KUBE_DNS_SVC_IP}/32" -p udp -m udp --sport 53 -j NOTRACK >/dev/null 2>&1; then
+  iptables -w 60 -t raw -A OUTPUT -s "${KUBE_DNS_SVC_IP}/32" -p udp -m udp --sport 53 -j NOTRACK
 fi
-if iptables -w 600 -t raw -C PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK >/dev/null 2>&1 ; then
-  # Remove. Will be added later, in liveness probe
-  iptables -w 600 -t raw -D PREROUTING -d ${KUBE_DNS_SVC_IP}/32 -m socket --nowildcard -j NOTRACK >/dev/null 2>&1
+if iptables -w 60 -t raw -C PREROUTING -d "${KUBE_DNS_SVC_IP}/32" -m socket --nowildcard -j NOTRACK >/dev/null 2>&1 ; then
+  # Remove. Will be added later.
+  iptables -w 60 -t raw -D PREROUTING -d "${KUBE_DNS_SVC_IP}/32" -m socket --nowildcard -j NOTRACK >/dev/null 2>&1
 fi
```
Comment (zuzzas) [inline ee/modules/350-node-local-dns/images/coredns/start.sh:31]: Write `not-ready` here.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

**Review by nabokihms — APPROVED** [review]

**Pull request merged**